packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/entity_type/mod.rs

```diff
@@ -1,15 +1,120 @@
 mod resolve;
 
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    future::Future,
+    pin::Pin,
+};
+
 use async_trait::async_trait;
 use error_stack::{IntoReport, Result, ResultExt};
+use futures::{stream, FutureExt, StreamExt, TryStreamExt};
 use tokio_postgres::GenericClient;
-use type_system::EntityType;
+use type_system::{uri::VersionedUri, EntityType};
 
 use crate::{
-    ontology::{AccountId, PersistedOntologyIdentifier},
-    store::{AsClient, EntityTypeStore, InsertionError, PostgresStore, UpdateError},
+    ontology::{
+        AccountId, EntityTypeTree, PersistedDataType, PersistedEntityType, PersistedLinkType,
+        PersistedOntologyIdentifier, PersistedPropertyType,
+    },
+    store::{
+        crud::Read,
+        postgres::{context::PostgresContext, PersistedOntologyType},
+        query::Expression,
+        AsClient, EntityTypeStore, InsertionError, PostgresStore, QueryError, UpdateError,
+    },
 };
 
+impl<C: AsClient> PostgresStore<C> {
+    /// Internal method to read a [`PersistedEntityType`] into a [`HashMap`].
+    ///
+    /// This is used to recursively resolve a type, so the result can be reused.
+    #[expect(
+        clippy::too_many_arguments,
+        reason = "Output is created in place and needs to be reused later"
+    )]
+    pub(crate) fn get_entity_type_as_dependency<'a>(
+        &'a self,
+        entity_type_uri: VersionedUri,
+        data_type_references: &'a mut HashMap<VersionedUri, PersistedDataType>,
+        property_type_references: &'a mut HashMap<VersionedUri, PersistedPropertyType>,
+        link_type_references: &'a mut HashMap<VersionedUri, PersistedLinkType>,
+        entity_type_references: &'a mut HashMap<VersionedUri, PersistedEntityType>,
+        data_type_resolve_depth: u8,
+        property_type_resolve_depth: u8,
+        link_type_query_depth: u8,
+        entity_type_query_depth: u8,
+    ) -> Pin<Box<dyn Future<Output = Result<(), QueryError>> + Send + 'a>> {
+        async move {
+            // TODO: Avoid cloning of URI
+            //   see https://stackoverflow.com/questions/51542024
```

Comment:
I think we shouldn't mark these as TODOs. They're misleading and prone to micro-optimisation.

I think it's fine for us to have something like 
```suggestion
            // URI is cloned due to limitations of Entry API, see 
            // https://stackoverflow.com/questions/51542024
```